atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java

```diff
@@ -2551,6 +2567,185 @@ public void exceptionThrownWhenTooManyPostFilterIterationsOccur() {
+                ImmutableMap.of(TABLE, Optional.of(ConflictHandler.SERIALIZABLE_LOCK_LEVEL_MIGRATION)));
+
+        txn.putWithMetadata(
+                TABLE, ImmutableMap.of(TEST_CELL, ValueAndChangeMetadata.of(TEST_VALUE, UPDATE_CHANGE_METADATA)));
+
+        LockDescriptor rowLock = AtlasRowLockDescriptor.of(TABLE.getQualifiedName(), TEST_CELL.getRowName());
+        LockDescriptor cellLock =
+                AtlasCellLockDescriptor.of(TABLE.getQualifiedName(), TEST_CELL.getRowName(), TEST_CELL.getColumnName());
+        verifyLockWasCalledWithLocksAndMetadataWhenCommitting(
+                txn,
+                timelockService,
+                LocksAndMetadata.of(
+                        ImmutableSet.of(rowLock, cellLock),
+                        Optional.of(LockRequestMetadata.of(
+                                ImmutableMap.of(rowLock, UPDATE_CHANGE_METADATA, cellLock, UPDATE_CHANGE_METADATA)))));
+    }
+
+    @Test
+    public void noMetadataResultsInEmptyOptional() {
+        TimelockService timelockService = spy(txManager.getTimelockService());
+        Transaction txn = getSnapshotTransactionWith(
+                timelockService, ImmutableMap.of(TABLE, Optional.of(ConflictHandler.SERIALIZABLE)));
+        txn.put(TABLE, ImmutableMap.of(TEST_CELL, TEST_VALUE));
+        verifyLockWasCalledWithLocksAndMetadataWhenCommitting(
+                txn,
+                timelockService,
+                LocksAndMetadata.of(
+                        ImmutableSet.of(AtlasRowLockDescriptor.of(TABLE.getQualifiedName(), TEST_CELL.getRowName())),
+                        Optional.empty()));
+    }
+
+    @Test
+    public void lockCollectionWithoutMetadataIsCorrectForRandomWrites() {
+        long randomSeed = System.currentTimeMillis();
+        int numTables = 10;
+        int numCellsPerTable = 100;
+        int totalWrites = 200;
+
+        Random random = new Random(randomSeed);
+
+        List<ConflictHandler> availableConflictHandlers = Arrays.asList(ConflictHandler.values());
+        List<TableReference> tables = IntStream.range(0, numTables)
+                .mapToObj(i -> {
+                    TableReference tableRef = TableReference.create(Namespace.create("random_test"), "table" + i);
+                    keyValueService.createTable(tableRef, AtlasDbConstants.GENERIC_TABLE_METADATA);
+                    return tableRef;
+                })
+                .collect(Collectors.toUnmodifiableList());
+        Map<TableReference, Optional<ConflictHandler>> tableToConflictHandler = KeyedStream.of(tables.stream())
+                .map(tableRef ->
+                        Optional.of(availableConflictHandlers.get(random.nextInt(availableConflictHandlers.size()))))
+                .collectToMap();
+        // Ensure that some rows have more than one cell
+        List<Cell> cells = IntStream.range(0, numCellsPerTable)
+                .mapToObj(i -> Cell.create(PtBytes.toBytes("row" + (i % 3)), PtBytes.toBytes("column" + i)))
```

Comment:
nit: we should pull out number of rows as well. I would also maybe being a bit more aggressive here, and doing a high number of rows, say 10?